File: Chimoney-Python/pychimoney/Info.py
from pychimoney import BaseAPI


class Info(BaseAPI):
    """
    This class handles the Info API
    """

    def airtime_countries(self):
        """
        This function returns a list of countries that supports airtime.
        """
        return self._handle_request("GET", "/v0.2/info/airtime-countries")

    def assets(self):
        """
        This function returns a list of supported assets.
        """
        return self._handle_request("GET", "/v0.2/info/assets")

    def banks(self, country="NG"):
        """
        This function returns a list of supported banks and banks code.
        """
        if not isinstance(country, str):
            raise TypeError("Country must be a string.")

        if not country:
            raise ValueError("Country is required.")

        return self._handle_request(
            "GET", "/v0.2/info/banks", params={"countryCode": country}
        )

    def currencies(self):
        """
        This function returns a list of supported currencies.
        """
        return self._handle_request("GET", "/v0.2/info/currencies")

    def mobile_money_codes(self):
        """
        This function returns a list of supported mobile money codes.
        """
        return self._handle_request("GET", "/v0.2/info/mobile-money-codes")

    def usd_to_local(self, destination_currency, ammount_in_usd):
        """
        This function returns the equivalent of USD in the destination
        currency.
        """

        if not isinstance(ammount_in_usd, int):
            raise TypeError("Ammount must be an integer.")

        if not isinstance(destination_currency, str):
            raise TypeError("Destination currency must be a string.")

        if not destination_currency and ammount_in_usd:
            raise ValueError("Destination currency and ammount are required.")

        return self._handle_request(
            "GET",
            "/v0.2/info/usd-to-local",
            params={
                "destinationCurrency": destination_currency,
                "amountInUSD": ammount_in_usd,
            },
        )

File: submissions/chimoney-js/modules/Info.js
const Joi = require("joi");
const { ValueError, TypeError } = require("../Errors");
const {
  handleRequest,
  HTTPMETHODS,
  formatJoiErrors,
} = require("../utils/helpers");

/**
 * This function returns a list of countries that support airtime
 * @returns The response from Chi Money API
 */
async function airtimeCountries() {
  return handleRequest({
    method: HTTPMETHODS.GET,
    path: "/v0.2/info/airtime-countries",
  });
}

/**
 * This function returns a list of supported assets
 * @returns The response from the Chi Money API
 */
async function assets() {
  return handleRequest({
    method: HTTPMETHODS.GET,
    path: "/v0.2/info/assets",
  });
}

/**
 *
 * @param {string} country The country code, default is Nigeria(NG).
 * @returns The response from Chi Money API
 */
async function banks(country = "NG") {
  // country is required
  if (!country) throw new ValueError("country is required");

  // country must be a string
  if (typeof country !== "string")
    throw new TypeError("country must be of type string");

  return handleRequest({
    method: HTTPMETHODS.GET,
    params: { countryCode: country },
    path: "/v0.2/info/country-banks",
  });
}

/**
 * This function returns the equivalent of local currency in USD
 * @param {string} originCurrency The source currency
 * @param {number} amountInOriginCurrency The amount in the origin currency
 * @returns The response from the Chi Money API
 */
async function localAmountInUSD(originCurrency, amountInOriginCurrency) {
  // Define validation schema
  const schema = Joi.object({
    originCurrency: Joi.string().required(),
    amountInOriginCurrency: Joi.number().required(),
  });

  // Validate input
  const { value, error } = schema.validate(
    {
      originCurrency,
      amountInOriginCurrency,
    },
    { abortEarly: false }
  );

  // Handle validation errors
  if (error) throw new ValueError("Invalid input(s)", formatJoiErrors(error));

  return handleRequest({
    method: HTTPMETHODS.GET,
    params: { ...value },
    path: "/v0.2/info/local-amount-in-usd",
  });
}

/**
 * This function returns a list of supported mobile money codes
 * @returns The response from the Chi Money API
 */
async function mobileMoneyCodes() {
  return handleRequest({
    method: HTTPMETHODS.GET,
    path: "/v0.2/info/mobile-money-codes",
  });
}

/**
 * This function returns the equivalent of USD in the destination currency.
 * @param {string} destinationCurrency The destination currency
 * @param {number} amountInUSD The amount in USD
 * @returns The response from the Chi Money API
 */
async function usdInLocalAmount(destinationCurrency, amountInUSD) {
  // Define validation schema
  const schema = Joi.object({
    destinationCurrency: Joi.string().required(),
    amountInUSD: Joi.number().required(),
  });

  // Validate input
  const { value, error } = schema.validate(
    {
      destinationCurrency,
      amountInUSD,
    },
    { abortEarly: false }
  );

  // Handle validation errors
  if (error) throw new ValueError("Invalid input(s)", formatJoiErrors(error));

  return handleRequest({
    method: HTTPMETHODS.GET,
    params: { ...value },
    path: "/v0.2/info/usd-amount-in-local",
  });
}

module.exports = {
  mobileMoneyCodes,
  usdInLocalAmount,
  localAmountInUSD,
  assets,
  airtimeCountries,
  banks,
};
